File: util/io/demux/callmap.go
package demux

import (
	"sync"

	"sigmaos/serr"
	sessp "sigmaos/session/proto"
)

// Map of outstanding calls indexed by tag
type callMap struct {
	sync.Mutex
	closed bool
	calls  map[sessp.Ttag]chan reply
}

func newCallMap() *callMap {
	return &callMap{calls: make(map[sessp.Ttag]chan reply)}
}

func (cm *callMap) close() error {
	cm.Lock()
	defer cm.Unlock()

	cm.closed = true
	return nil
}

func (cm *callMap) isClosed() bool {
	cm.Lock()
	defer cm.Unlock()

	return cm.closed
}

func (cm *callMap) put(tag sessp.Ttag, ch chan reply) *serr.Err {
	cm.Lock()
	defer cm.Unlock()

	if cm.closed {
		return serr.NewErr(serr.TErrUnreachable, "dmxclnt")
	}
	cm.calls[tag] = ch
	return nil
}

func (cm *callMap) remove(tag sessp.Ttag) (chan reply, bool) {
	cm.Lock()
	defer cm.Unlock()

	if ch, ok := cm.calls[tag]; ok {
		delete(cm.calls, tag)
		return ch, true
	}
	return nil, false
}

func (cm *callMap) outstanding() []sessp.Ttag {
	cm.Lock()
	defer cm.Unlock()

	ts := make([]sessp.Ttag, 0, len(cm.calls))
	for t, _ := range cm.calls {
		ts = append(ts, t)
	}
	return ts
}

File: cpp/io/demux/internal/callmap.cc
#include <io/demux/internal/callmap.h>

namespace sigmaos {
namespace io::demux::internal {

std::expected<int, sigmaos::serr::Error> CallMap::Put(uint64_t seqno, std::unique_ptr<std::promise<std::expected<std::shared_ptr<sigmaos::io::transport::Call>, sigmaos::serr::Error>>> p) {
  std::lock_guard<std::mutex> guard(_mu);
  if (_closed) {
    return std::unexpected(sigmaos::serr::Error(sigmaos::serr::TErrUnreachable, "Err: demux closed"));
  }
  _calls[seqno] = std::move(p);
  return 0;
}

std::optional<std::unique_ptr<std::promise<std::expected<std::shared_ptr<sigmaos::io::transport::Call>, sigmaos::serr::Error>>>> CallMap::Remove(uint64_t seqno) {
  if (!_calls.contains(seqno)) {
    return std::nullopt;
  }
  auto call = std::move(_calls.extract(seqno).mapped());
  return call;
}

std::vector<uint64_t> CallMap::Outstanding() {
  std::vector<uint64_t> outstanding(_calls.size());
  int i = 0;
  for (const auto &pair: _calls) {
    outstanding[i] = pair.first;
  }
  return outstanding;
}

void CallMap::Close() {
  std::lock_guard<std::mutex> guard(_mu);
  _closed = true;
}

bool CallMap::IsClosed() {
  std::lock_guard<std::mutex> guard(_mu);
  return _closed;
}

};
};
